### DISCUSSION TITLE
How to create an extension?

### DISCUSSION THREAD

**Discussion by Moosems** [whmsft/WhirlEdit#9] [Q&A] [answered]
How could I create an extension?

**Comment by ayuvius** [comment, marked as answer]
## tiny documentation on extensions

I haven't worked enough extensions BTW.

Extensions are just codes executed with an `exec()` function..

### location
Extensions are saved in new folders made in `DATA/extensions` and main codes are in `__init__.py` file in that directory..

### Manager file's structure

There is a class called `tasks` which has 12 lists namely:

```
1. after_imports (when modules are imported)
2. before_extension_installation (before extensions are loaded)
3. after_extension_installation (after extensions are loaded)
4. main_vars_definition (with important variables definition)
5. main_funcs_definition (with main funcs definition) 
6. main_classes_definition (with important classes' definition) 
7. before_root_definition (before main window definition) 
8. after_root_definition (after main window definition) 
9. sidebar_widgets (after sidebar widgets: files, runner, customizer and settings definition)
10. before_mainloop (before root.mainloop)
11. before_configs_save (after root.mainloop)
12. onexit (just before exiting)
```

Each contains codes to be executed in certain places of main program

### Code

Extensions' codes are in form of multiline strings example:

```python
tasks.before_mainloop.append("""
Menubar = Menu(root, activebackground="#0084FF", activeforeground="#FFFFFF",
               bg="#FFFFFF", fg="#0084FF", font="consolas")
Filemenu = Menu(root, tearoff = 0)
Menubar.add_cascade(label="File", menu=Filemenu)
```

This creates a file menu in the menubar.. 

I also wrote a "just-working" title bar (menu bar) earlier at [this link](https://github.com/whmsft/WhirlEdit/blob/bcaba1a7f97f10d72b272bf18dcff8a459e08b90/DATA/extensions/menubar/__init__.py)

### Final words

This sounds hectic as you need to read that 1400+ lines of code before making extensions thus I may rewrite that entirely..
